libsolutil/StringUtils.h

```diff
@@ -101,109 +101,109 @@ std::string joinHumanReadablePrefixed
 		return _separator + joinHumanReadable(_list, _separator, _lastSeparator);
 }
 
-/// Same as @ref formatNumberReadable but only for unsigned numbers
+/// Formats large numbers to be easily readable by humans.
+/// Returns decimal representation for smaller numbers; hex for large numbers.
+/// "Special" numbers, powers-of-two and powers-of-two minus 1, are returned in
+/// formulaic form like 0x01 * 2**24 - 1.
+/// @a T can be any integer type, will typically be u160, u256 or bigint.
+/// @param _value to be formatted
+/// @param _useTruncation if true, internal truncation is also applied,
+/// like  0x5555...{+56 more}...5555
+/// @example formatNumberReadable((u256)0x7ffffff) = "2**27 - 1"
+/// @example formatNumberReadable(-57896044618658097711785492504343953926634992332820282019728792003956564819968) = -2**255
 template <class T>
-inline std::string formatUnsignedNumberReadable (
+inline std::string formatNumberReadable(
 	T const& _value,
 	bool _useTruncation = false
 )
 {
 	static_assert(
-		std::is_same<bigint, T>::value || !std::numeric_limits<T>::is_signed,
-		"only unsigned types or bigint supported"
-	); //bigint does not carry sign bit on shift
+		std::numeric_limits<T>::is_integer,
+		"only integer types are supported"
+	);
+
+	bool isNegative = _value < 0;
+	bigint temp;
+	bigint signedValue = isNegative ? (bigint(-1) * _value) : bigint(_value);
+	std::string sign = isNegative ? "-" : "";
 
 	// smaller numbers return as decimal
-	if (_value <= 0x1000000)
-		return _value.str();
+	if (signedValue <= 0x1000000)
+		return sign + signedValue.str();
 
 	HexCase hexcase = HexCase::Mixed;
-	HexPrefix prefix = HexPrefix::Add;
+	HexPrefix hexprefix = HexPrefix::Add;
 
 	// when multiple trailing zero bytes, format as N * 2**x
 	int i = 0;
-	T v = _value;
-	for (; (v & 0xff) == 0; v >>= 8)
+	for (temp = signedValue; (temp & 0xff) == 0; temp >>= 8)
```

Comment:
I think it may be better scoping `temp` here and in the next loop, as they are not used inbetween.

Reply:
Would also keep it named `v` as before.

Reply:
wouldn't using `temp` here make it more clear what the role of the variable is? It took me some time to understand what role `v` played in the code; and I think other programmers would face a similar situation when looking at the code.

Reply:
I agree with naming the variable as `temp` and also with limiting its scope to only inside the loops.

Reply:
ok, will do that